internal/rust/semantic/src/types.rs

```diff
@@ -0,0 +1,705 @@
+        match self {
+            Row::Empty => f.write_str("{}"),
+            Row::Var(tv) => write!(f, "{}", tv),
+            Row::Extension { head: h, tail: t } => {
+                if let Err(e) = write!(f, "{} | ", h) {
+                    return Err(e);
+                }
+                t.display(f)
+            }
+        }
+    }
+
+    // Flatten a record type into a hashmap of property names and types
+    fn flatten(&self, props: &mut HashMap<String, MonoType>) -> Option<Tvar> {
+        match self {
+            Row::Empty => None,
+            Row::Var(tv) => Some(*tv),
+            Row::Extension { head: h, tail: t } => {
+                props.insert(h.k.clone(), h.v.clone());
+                t.flatten(props)
+            }
+        }
+    }
+}
+
+// A key value pair representing a property type in a record
+#[derive(Debug, Clone, PartialEq)]
+pub struct Property {
+    pub k: String,
+    pub v: MonoType,
+}
+
+impl fmt::Display for Property {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}:{}", self.k, self.v)
+    }
+}
+
+// Function represents a function type.
+//
+// A function type is defined by as set of required arguments,
+// a set of optional arguments, an optional pipe argument, and
+// a required return type.
+//
+#[derive(Debug, Clone, PartialEq)]
+pub struct Function {
+    pub req: HashMap<String, MonoType>,
+    pub opt: HashMap<String, MonoType>,
+    pub pipe: Option<Property>,
+    pub retn: MonoType,
+}
+
+impl fmt::Display for Function {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut req: Vec<Property> = Vec::new();
+        for (k, v) in &self.req {
+            req.push(Property {
+                k: k.clone(),
+                v: v.clone(),
+            });
+        }
+        req.sort_unstable_by(|a, b| a.k.cmp(&b.k));
+
+        let mut opt: Vec<Property> = Vec::new();
+        for (k, v) in &self.opt {
+            opt.push(Property {
+                k: String::from("?") + &k,
+                v: v.clone(),
+            });
+        }
+        opt.sort_unstable_by(|a, b| a.k.cmp(&b.k));
+
+        let mut args: Vec<Property> = Vec::new();
+        if let Some(pipe) = &self.pipe {
+            if pipe.k == "<-" {
+                args.push(pipe.clone());
+            } else {
+                args.push(Property {
+                    k: String::from("<-") + &pipe.k,
+                    v: pipe.v.clone(),
+                });
+            }
+        }
+
+        args.append(&mut req);
+        args.append(&mut opt);
+        write!(
+            f,
+            "({}) -> {}",
+            DisplayList {
+                values: &args,
+                delim: ", "
+            },
+            self.retn
+        )
+    }
+}
+
+// BoundTvar represents a constrained type variable.
+// Used solely for displaying the generic constraints of a polytype.
+#[derive(Debug)]
+struct BoundTvar {
```

Comment:
nit, BoundTvar made me think of the type variable being bound in the environment not as being constrained by a kind. Maybe a name like TvarKind or TKPair or something like that that communicates is constrained by a kind would be more clear? 

Its also fine as BoundTvar since the comment makes it clear. Do what ever you think works.